module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,182 @@
+	header, err := i.headers.ByBlockID(ID)
+	if err != nil {
+		return 0, fmt.Errorf("could not find block by ID %: %w", ID, err)
+	}
+
+	return header.Height, nil
+}
+
+// Commitment retrieves a commitment at the provided block height.
+// If a commitment at height is not found expect a `storage.ErrNotFound` error.
+func (i *ExecutionState) Commitment(height uint64) (flow.StateCommitment, error) {
+	// todo use the commitment storage to read heights
+	//if height < i.startIndexHeight || height > i.lastIndexedHeight.Value() {
+	//	return flow.DummyStateCommitment, fmt.Errorf(
+	//		"state commitment out of indexed height bounds, current height range: [%d, %d], requested height: %d",
+	//		i.startIndexHeight,
+	//		i.lastIndexedHeight.Value(),
+	//		height,
+	//	)
+	//}
+
+	val, ok := i.commitments[height]
+	if !ok {
+		return flow.DummyStateCommitment, fmt.Errorf("could not find commitment at height %d", height)
+	}
+
+	return val, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// If the register was not found the storage.ErrNotFound error is returned.
+func (i *ExecutionState) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	first, err := i.registers.FirstHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read first index height: %w", err)
+	}
+
+	last, err := i.registers.LatestHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read last index height: %w", err)
+	}
+
+	if height < first || height > last {
+		return nil, fmt.Errorf("register out of indexed height bounds, current height range: [%d, %d], requested height: %d", first, last, height)
+	}
+
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := i.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// If the height was already indexed the operation will be ignored.
+func (i *ExecutionState) IndexBlockData(ctx context.Context, data *execution_data.BlockExecutionDataEntity) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	block, err := i.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	// progress height in storage first, if the height is incorrect the index storage will panic
+	// this is due to unrecoverable bug in the implementation
+	err = i.registers.SetLatestHeight(block.Height)
+	if err != nil {
+		return fmt.Errorf("failed to progress index height to %d: %w", block.Height, err)
+	}
+
+	// concurrently process indexing of block data
+	g, ctx := errgroup.WithContext(ctx)
+	for j, chunk := range data.ChunkExecutionDatas {
```

Comment:
note: the last `chunk` is for the system chunk. this should be handled differently:
* The collection should not be indexed. It is auto generated as needed.
* All events should be indexed as `ServiceEvent` instead of standard user `Event`